kernel/posix/pthread.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <atomic.h>
+#include "ksched.h"
+#include "wait_q.h"
+
+/* Pthread structure for number of threads */
+struct posix_thread pthread[CONFIG_MAX_PTHREAD_COUNT];
+
+static bool is_posix_prio_valid(u32_t priority, int policy)
+{
+	if ((priority >= sched_get_priority_min(policy)) &&
+	    (priority <= sched_get_priority_max(policy))) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline s32_t posix_to_zephyr_priority(u32_t priority, int policy)
+{
+	s32_t prio;
+
+	if (policy == SCHED_FIFO) {
+		/* Zephyr COOP priority starts from -1 */
+		prio =  -1 * (priority + 1);
+	} else {
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - priority);
+	}
+
+	return prio;
+}
+
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	if (is_posix_prio_valid(priority, attr->schedpolicy) == false) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	attr->priority = priority;
+	return 0;
+}
+
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+	if (stackaddr == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
+}
+
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	int thread_index = 0;
```

Comment:
Use an `atomic_val_t` here: that's what `atomic_inc()` returns.  Even though it's just a typedef to int, it's better to always use the correct type to future-proof the code, readability, and avoid gotchas with implicit integer conversions/losing precision.